### PR TITLE
TIP-255: Improve FilterRegistry to find on supported operators

### PR DESCRIPTION
**Pull request by willy-ahva**
| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | y
| Added Behats                      | n
| Changelog updated                 | y
| Review and 2 GTM                  |
| Micro Demo to the PO (Story only) |
| Migration script                  |
| Tech Doc                          |

### CODE REVIEW

File: CHANGELOG-1.6.md [hunk 1]
```diff
@@ -43,6 +43,7 @@
 - PIM-5657: Introduce a `JobTemplateProvider` that holds the job template codes to use for creating, showing, editing job profiles. The provider uses configuration files in order to retrieve overridden templates for specific job names
 - TIP-458: Move the Converters from Processors to Readers. Now, all the readers return a standard format as output, and all the processors get a standard format as input.
 - TIP-459: Standardize the denormalization Processors, to use SimpleProcessor in most of the cases.
+- TIP-255: Allow to select filter on supported operator, to add new operators easily on existing field/attribute type
```
Comment (phaseinducer) [inline CHANGELOG-1.6.md]: Maybe specify it is for PQB filters

File: src/Pim/Bundle/CatalogBundle/Command/ProductQueryHelp/AttributeFilterDumper.php [hunk 2]
```diff
@@ -39,30 +40,55 @@ public function __construct(FilterRegistryInterface $registry, AttributeReposito
...
+     */
+    protected function getAttributeFilters()
+    {
+        $attributeFilters = [];
+        foreach ($this->registry->getAttributeFilters() as $filter) {
+            $supportedAttributes = $filter->getAttributes();
+
+            if (null !== $supportedAttributes) {
+                foreach ($supportedAttributes as $attribute) {
+                    $attributeFilters[$attribute] = $filter;
```
Comment (phaseinducer) [inline src/Pim/Bundle/CatalogBundle/Command/ProductQueryHelp/AttributeFilterDumper.php]: How can it work if several filters support the same attribute?

File: src/Pim/Bundle/CatalogBundle/Command/ProductQueryHelp/AttributeFilterDumper.php [hunk 3]
```diff
@@ -39,30 +40,55 @@ public function __construct(FilterRegistryInterface $registry, AttributeReposito
...
+                }
+            } else {
```
Comment (phaseinducer) [inline src/Pim/Bundle/CatalogBundle/Command/ProductQueryHelp/AttributeFilterDumper.php]: I don't understand this part.

File: src/Pim/Bundle/CatalogBundle/Command/ProductQueryHelp/AttributeFilterDumper.php (hunk 2 above)
Comment (phaseinducer) [inline src/Pim/Bundle/CatalogBundle/Command/ProductQueryHelp/AttributeFilterDumper.php]: Bad naming imho `$attribute` -> `$attributeType`

File: src/Pim/Bundle/CatalogBundle/Command/ProductQueryHelp/FieldFilterDumper.php [hunk 4]
```diff
@@ -33,33 +33,15 @@ public function __construct(FilterRegistryInterface $registry)
...
-            }
-            $rows[] = [$field, $class, $operators];
+        foreach ($this->registry->getFieldFilters() as $filter) {
+            $class     = get_class($filter);
+            $operators = implode(', ', $filter->getOperators());
+            $fields    = implode(', ', $filter->getFields());
+            $rows[]    = [$class, $fields, $operators];
         }
-        $headers = ['field', 'filter_class', 'operators'];
+        $headers = ['filter_class', 'fields', 'operators'];
```
Comment (phaseinducer) [inline src/Pim/Bundle/CatalogBundle/Command/ProductQueryHelp/FieldFilterDumper.php]: Why switch this?

File: src/Pim/Bundle/CatalogBundle/Doctrine/Common/Filter/CategoryFilter.php [hunk 5]
```diff
@@ -118,6 +118,14 @@ public function supportsField($field)
     /**
      * {@inheritdoc}
      */
+    public function getFields()
```
Comment (phaseinducer) [inline src/Pim/Bundle/CatalogBundle/Doctrine/Common/Filter/CategoryFilter.php]: getSupportedFields() ?

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Filter/AbstractAttributeFilter.php [hunk 6]
```diff
@@ -19,6 +19,25 @@
     /** @var AttributeValidatorHelper */
     protected $attrValidatorHelper;
 
+    /** @var string[] */
+    protected $supportedAttributes;
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getAttributes()
```
Comment (phaseinducer) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Filter/AbstractAttributeFilter.php]: getSupportedAttributes() ?

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Filter/BooleanFilter.php [hunk 7]
```diff
@@ -54,9 +51,9 @@ public function supportsField($field)
     /**
      * {@inheritdoc}
      */
-    public function supportsAttribute(AttributeInterface $attribute)
+    public function getFields()
```
Comment (phaseinducer) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Filter/BooleanFilter.php]: Why don't put it in `AbstractAttributeFilter`?

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Filter/MediaFilter.php [hunk 8]
```diff
@@ -117,7 +106,7 @@ protected function prepareValue($operator, $value)
      * @param AttributeInterface $attribute
      * @param mixed              $value
      */
-    public function checkValue(AttributeInterface $attribute, $value)
+    protected function checkValue(AttributeInterface $attribute, $value)
```
Comment (phaseinducer) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Filter/MediaFilter.php]: :+1:

File: src/Pim/Bundle/CatalogBundle/Doctrine/Common/Filter/CategoryFilter.php [hunk 9]
```diff
@@ -37,7 +37,7 @@ class CategoryFilter implements FieldFilterInterface
     protected $supportedOperators;
 
     /**
-     * Instanciate the base filter
+     * Instantiate the base filter
```
Comment (phaseinducer) [inline src/Pim/Bundle/CatalogBundle/Doctrine/Common/Filter/CategoryFilter.php]: You can drop this line

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Filter/ProductIdFilter.php [hunk 10]
```diff
@@ -19,7 +19,7 @@ class ProductIdFilter extends AbstractFilter implements FieldFilterInterface
     protected $supportedFields;
 
     /**
-     * Instanciate the filter
+     * Instantiate the filter
```
Comment (phaseinducer) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Filter/ProductIdFilter.php]: Remove it

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Filter/BooleanFilter.php [hunk 11]
```diff
@@ -19,13 +19,10 @@
...
 {
     /** @var array */
-    protected $supportedAttributes;
-
-    /** @var array */
     protected $supportedFields;
 
     /**
-     * Instanciate the base filter
+     * Instantiate the base filter
```
Comment (phaseinducer) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Filter/BooleanFilter.php]: idem

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Filter/NumberFilter.php [hunk 12]
```diff
@@ -17,11 +17,8 @@
  */
 class NumberFilter extends AbstractAttributeFilter implements AttributeFilterInterface
 {
-    /** @var array */
-    protected $supportedAttributes;
-
     /**
-     * Instanciate the base filter
+     * Instantiate the base filter
```
Comment (phaseinducer) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Filter/NumberFilter.php]: :fire:

File: src/Pim/Component/Catalog/Query/Filter/FilterRegistryInterface.php [hunk 13]
```diff
@@ -23,27 +23,44 @@ public function register(FilterInterface $filter);
...
+     * Returns all field filters
+     *
+     * @return FieldFilterInterface[]|null
+     */
+    public function getFieldFilters();
+
+    /**
+     * Returns all attribute filters
+     *
+     * @return AttributeFilterInterface[]
```
Comment (phaseinducer) [inline src/Pim/Component/Catalog/Query/Filter/FilterRegistryInterface.php]: `|null`?

File: src/Pim/Bundle/CatalogBundle/Doctrine/Common/Filter/CategoryFilter.php (hunk 5 above)
Comment (willy-ahva) [inline src/Pim/Bundle/CatalogBundle/Doctrine/Common/Filter/CategoryFilter.php]: We have a getOperators(). It is to be consistent.

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Filter/AbstractAttributeFilter.php (hunk 6 above)
Comment (willy-ahva) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Filter/AbstractAttributeFilter.php]: same

File: src/Pim/Component/Catalog/Query/Filter/FilterRegistryInterface.php (hunk 13 above)
Comment (willy-ahva) [inline src/Pim/Component/Catalog/Query/Filter/FilterRegistryInterface.php]: hum in fact it is always an array. So I'll remove |null from `getFieldFilters()`

File: src/Pim/Bundle/CatalogBundle/Command/ProductQueryHelp/FieldFilterDumper.php (hunk 4 above)
Comment (willy-ahva) [inline src/Pim/Bundle/CatalogBundle/Command/ProductQueryHelp/FieldFilterDumper.php]: It is indexed by filter_class now. I have all the same information but not in the same order. That was easier for me as I does not receive information in the same state that before.

File: src/Pim/Bundle/CatalogBundle/Command/ProductQueryHelp/AttributeFilterDumper.php [hunk 14]
```diff
@@ -39,30 +41,102 @@ public function __construct(FilterRegistryInterface $registry, AttributeReposito
     public function dump(OutputInterface $output, HelperSet $helperSet)
     {
         $output->writeln("<info>Useable attributes filters...</info>");
-        $attributes = $this->repository->findAll();
+        $attributeFilters = $this->getAttributeFilters();
+        $attributes       = $this->repository->findAll();
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Command/ProductQueryHelp/AttributeFilterDumper.php]: no way josé ;)

File: src/Pim/Bundle/CatalogBundle/Command/ProductQueryHelp/AttributeFilterDumper.php [hunk 15]
```diff
@@ -39,30 +41,102 @@ public function __construct(FilterRegistryInterface $registry, AttributeReposito
...
+        return $attributeFilters;
+    }
+
+    /**
+     * @param AttributeInterface $attribute
+     * @param array              $attributeFilters
+     *
+     * @return array
+     */
+    protected function generateNewEntries(AttributeInterface $attribute, array $attributeFilters)
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Command/ProductQueryHelp/AttributeFilterDumper.php]: I don't understand at all this method

File: src/Pim/Bundle/CatalogBundle/Command/ProductQueryHelp/FieldFilterDumper.php (hunk 4 above)
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Command/ProductQueryHelp/FieldFilterDumper.php]: it's not logical
as a dev, you look for a filter for a particuler field

I disagree with this change :/

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Filter/BooleanFilter.php (hunk 7 above)
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Filter/BooleanFilter.php]: why not in the abstract?

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Filter/CompletenessFilter.php [hunk 16]
```diff
@@ -51,6 +51,14 @@ public function supportsField($field)
 
     /**
      * {@inheritdoc}
+     */
+    public function getFields()
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Filter/CompletenessFilter.php]: why not a AbstractFieldFilter?

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Filter/DateTimeFilter.php [hunk 17]
```diff
@@ -44,6 +44,14 @@ public function supportsField($field)
     /**
      * {@inheritdoc}
      */
+    public function getFields()
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Filter/DateTimeFilter.php]: abstractFieldFilter?

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Filter/FamilyFilter.php [hunk 18]
```diff
@@ -48,6 +48,14 @@ public function supportsField($field)
     /**
      * {@inheritdoc}
      */
+    public function getFields()
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Filter/FamilyFilter.php]: /home/jjanvier/akeneo/pim_master/pcd

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Filter/GroupsFilter.php [hunk 19]
```diff
@@ -48,6 +48,14 @@ public function supportsField($field)
     /**
      * {@inheritdoc}
      */
+    public function getFields()
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Filter/GroupsFilter.php]: abstractFieldFilter?

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Filter/AbstractAttributeFilter.php [hunk 20]
```diff
@@ -19,6 +19,25 @@
     /** @var AttributeValidatorHelper */
     protected $attrValidatorHelper;
 
+    /** @var string[] */
+    protected $supportedAttributes;
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Filter/AbstractAttributeFilter.php]: supportedAttributeTypes?

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Filter/AbstractAttributeFilter.php [hunk 21]
```diff
@@ -19,6 +19,25 @@
...
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getAttributes()
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Filter/AbstractAttributeFilter.php]: getAttributeTypes?

File: src/Pim/Component/Catalog/Query/ProductQueryBuilder.php [hunk 22]
```diff
@@ -108,20 +108,14 @@ public function addFilter($field, $operator, $value, array $context = [])
...
         if (null === $filter) {
             throw new \LogicException(
-                sprintf('Filter on field "%s" is not supported', $field)
-            );
-        }
-
-        if (!$filter->supportsOperator($operator)) {
-            throw new \LogicException(
-                sprintf('Filter on field "%s" doesn\'t provide operator "%s"', $field, $operator)
+                sprintf('Filter on field "%s" is not supported or does not support operator "%s"', $field, $operator)
```
Comment (jjanvier) [inline src/Pim/Component/Catalog/Query/ProductQueryBuilder.php]: property instead of field ;)

**Comment by jjanvier** [comment]
except the dumper part that I don't understand at all, GTM

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Filter/DateFilter.php [hunk 23]
```diff
@@ -41,14 +38,6 @@ public function __construct(
     /**
      * {@inheritdoc}
      */
-    public function supportsAttribute(AttributeInterface $attribute)
```
Comment (solivier) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Filter/DateFilter.php]: bc break

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Filter/MediaFilter.php [hunk 24]
```diff
@@ -39,14 +36,6 @@ public function __construct(
     /**
      * {@inheritdoc}
      */
-    public function supportsAttribute(AttributeInterface $attribute)
```
Comment (solivier) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Filter/MediaFilter.php]: bc break

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Filter/MetricFilter.php [hunk 25]
```diff
@@ -53,14 +50,6 @@ public function __construct(
     /**
      * {@inheritdoc}
      */
-    public function supportsAttribute(AttributeInterface $attribute)
```
Comment (solivier) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Filter/MetricFilter.php]: bc break

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Filter/BooleanFilter.php [hunk 26]
```diff
@@ -100,16 +95,16 @@ public function addFieldFilter($field, $operator, $value, $locale = null, $scope
...
     {
-        return in_array($field, $this->supportedFields);
+        return $this->supportedFields;
     }
 
     /**
      * {@inheritdoc}
      */
-    public function supportsAttribute(AttributeInterface $attribute)
+    public function supportsField($field)
```
Comment (solivier) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Filter/BooleanFilter.php]: bc break

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Filter/DateFilter.php (hunk 23 above)
Comment (willy-ahva) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Filter/DateFilter.php]: Nope, this method has been moved in Abstract so no BC break

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Filter/MediaFilter.php (hunk 24 above)
Comment (willy-ahva) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Filter/MediaFilter.php]: same

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Filter/MetricFilter.php (hunk 25 above)
Comment (willy-ahva) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Filter/MetricFilter.php]: same

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Filter/BooleanFilter.php (hunk 26 above)
Comment (willy-ahva) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Filter/BooleanFilter.php]: same

File: CHANGELOG-1.6.md [hunk 27]
```diff
@@ -346,3 +347,7 @@
 - Add parameter `$operationGroup` to `Pim\Bundle\EnrichBundle\MassEditAction\MassEditFormResolver::getAvailableOperationsForm()`
 - Add parameter `$operationGroup` to `Pim\Bundle\EnrichBundle\MassEditAction\Operation\OperationRegistryInterface::register()`
 - Replace method `getAllByGridName()` by `getAllByGridNameAndGroup()` in `Pim\Bundle\EnrichBundle\MassEditAction\Operation\OperationRegistryInterface`
+- Change visibility of `checkValue()` method of `Pim\Bundle\CatalogBundle\Doctrine\ORM\Filter\MediaFilter` from public to protected
+- Add `getAttribute()` method to `Pim\Component\Catalog\Query\Filter\AttributeFilterInterface`
```
Comment (phaseinducer) [inline CHANGELOG-1.6.md]: getAttributeTypes

File: src/Pim/Bundle/CatalogBundle/Command/ProductQueryHelp/AttributeFilterDumper.php [hunk 28]
```diff
@@ -39,30 +41,104 @@ public function __construct(FilterRegistryInterface $registry, AttributeReposito
...
-                $class,
-                $operators
-            ];
+            $rows = array_merge($rows, $this->getFilterInformationForAttribute($attribute, $attributeFilters));
         }
-        $table = $helperSet->get('table');
-        $headers = ['attribute', 'localizable', 'scopable', 'attribute type', 'filter_class', 'operators'];
+
+        $table   = $helperSet->get('table');
+        $headers = ['attribute', 'localizable', 'scopable', 'attribute type', 'operators', 'filter_class'];
```
Comment (phaseinducer) [inline src/Pim/Bundle/CatalogBundle/Command/ProductQueryHelp/AttributeFilterDumper.php]: These 3 lines could be:
```
$table
    ->setHeaders(['attribute', 'localizable', 'scopable', 'attribute type', 'operators', 'filter_class'])
    ->setRows($rows)
    ->render($output)
;
```

**Comment by phaseinducer** [comment]
GTM

**Pull request merged**